### PR TITLE
libbpf-rs: Add OpenMap::reuse_map() API

### LINKED ISSUE #52: Interacting with existing maps?
Long story short: I want to use libbf-rs to interact with globally pinned maps that are not created by loading an object with libbpf-rs, but I can not find a way to do that.

Some context: I am working on a user space program which communicates with a bpf object loaded as a tc direct-action via bpf maps. My bpf object in kernel creates the maps, and pins them globally. So, the task for the user-space part is just to manipulate these existing maps. I came up with this design since I have not find any option to load my bpf object as a direct-action w/o using tc. 

Libbpf has a straightforward way to interact with existing maps: bpf_obj_get() returns the fd of the map, then the returned fd can be used for map operations.  Is there any way to do this with libbpf-rs?

Thanks!

### PR DESCRIPTION
**Pull request by danobi**
Useful for interacting with pinned maps.

This closes #52.

### CODE REVIEW

File: libbpf-rs/src/map.rs [hunk 1]
```diff
@@ -59,6 +59,23 @@ impl OpenMap {
     pub fn set_inner_map_fd(&mut self, inner: &Map) {
         unsafe { libbpf_sys::bpf_map__set_inner_map_fd(self.ptr, inner.fd()) };
     }
+
+    /// Reuse a already-pinned map for `self`.
```
Comment (anakryiko) [inline libbpf-rs/src/map.rs:63]: typo: a -> an

File: libbpf-rs/src/map.rs [hunk 2]
```diff
@@ -59,6 +59,23 @@ impl OpenMap {
...
+    pub fn reuse_map<P: AsRef<Path>>(&mut self, path: P) -> Result<()> {
```
Comment (anakryiko) [inline libbpf-rs/src/map.rs:64]: I wonder if `reuse_pinned_map` would be a clearer name?

File: libbpf-rs/tests/test.rs [hunk 3]
```diff
@@ -242,3 +242,57 @@ fn test_object_link_pin() {
...
+    let obj_path = get_test_object_path();
+    let mut builder = ObjectBuilder::default();
+    builder.debug(true);
+    let mut open_obj = builder.open_file(obj_path).expect("failed to open object");
+
+    let start = open_obj
+        .map("start")
+        .expect("error finding map")
+        .expect("failed to find map");
+    assert!(start.reuse_map("/asdf").is_err());
```
Comment (anakryiko) [inline libbpf-rs/tests/test.rs:280]: will assert!() panic? if yes, you won't unpin the map, wouldn't that be a problem?

File: libbpf-rs/src/map.rs [hunk 4]
```diff
@@ -59,6 +59,23 @@ impl OpenMap {
...
+        let cstring = util::path_to_cstring(path)?;
+
+        let fd = unsafe { libbpf_sys::bpf_obj_get(cstring.as_ptr()) };
+        if fd < 0 {
+            return Err(Error::System(errno::errno()));
+        }
+
+        let ret = unsafe { libbpf_sys::bpf_map__reuse_fd(self.ptr, fd) };
```
Comment (anakryiko) [inline libbpf-rs/src/map.rs:72]: looking at the code of bpf_map__reuse_fd(), I think you are supposed to close(fd) whether this call succeeded or failed.

**Review by anakryiko — CHANGES_REQUESTED** [review]

File: libbpf-rs/src/map.rs (hunk 4 above)
Comment (danobi) [reply libbpf-rs/src/map.rs:72]: Funny story: I woke up in the middle of the night and thought "I forgot to close the fd" and when I woke up this morning I forgot until now

File: libbpf-rs/tests/test.rs (hunk 3 above)
Comment (danobi) [reply libbpf-rs/tests/test.rs:280]: Good point; I'll fix it

File: libbpf-rs/src/map.rs [hunk 5]
```diff
@@ -59,6 +59,35 @@ impl OpenMap {
...
+        // Always close `fd` regardless of if `bpf_map__reuse_fd` succeeded or failed
+        match unistd::close(fd) {
+            Ok(_) => (),
+            Err(e) => {
+                return Err(Error::Internal(format!(
+                    "Failed to close pinned map fd: {}",
+                    e
+                )))
+            }
+        }
```
Comment (anakryiko) [inline libbpf-rs/src/map.rs:83]: there isn't much you can do if close() fails, and in practice it shouldn't ever fail, so I'd drop this error handling part

**Review by anakryiko — APPROVED** [review]
defer! {} is nice!

**Pull request merged**